linera-core/src/client/validator_manager/manager.rs

```diff
@@ -0,0 +1,1268 @@
+        // Create a manager with a very short in-flight timeout
+        let manager = create_test_manager(Duration::from_millis(50), Duration::from_secs(60));
+
+        let key = test_key();
+
+        // Track how many times the operation is executed
+        let execution_count = Arc::new(AtomicUsize::new(0));
+
+        // Create a channel to control when the first operation completes
+        let (tx, rx) = oneshot::channel();
+
+        // Start first request (will be slow - waits for signal)
+        let manager_clone = Arc::clone(&manager);
+        let key_clone = key.clone();
+        let execution_count_clone = execution_count.clone();
+        let first_request = tokio::spawn(async move {
+            manager_clone
+                .deduplicated_request(key_clone, (), |_| async move {
+                    execution_count_clone.fetch_add(1, Ordering::SeqCst);
+                    rx.await.unwrap();
+                    test_result_ok()
+                })
+                .await
+        });
+
+        // Give the first request time to register as in-flight
+        tokio::time::sleep(Duration::from_millis(10)).await;
+
+        // Wait for the timeout to elapse
+        tokio::time::sleep(Duration::from_millis(MAX_REQUEST_TTL_MS + 1)).await;
+
+        // Start second request - should NOT deduplicate because first request exceeded timeout
+        let execution_count_clone2 = execution_count.clone();
+        let second_request = tokio::spawn(async move {
+            manager
+                .deduplicated_request(key, (), |_| async move {
+                    execution_count_clone2.fetch_add(1, Ordering::SeqCst);
+                    test_result_ok()
+                })
+                .await
+        });
+
+        // Wait for second request to complete
+        let result2: Result<Vec<ConfirmedBlockCertificate>, NodeError> =
+            second_request.await.unwrap();
+        assert!(result2.is_ok());
+
+        // Complete the first request
+        tx.send(()).unwrap();
+        let result1: Result<Vec<ConfirmedBlockCertificate>, NodeError> =
+            first_request.await.unwrap();
+        assert!(result1.is_ok());
+
+        // Operation should have been executed twice (timeout triggered new request)
+        assert_eq!(execution_count.load(Ordering::SeqCst), 2);
+    }
+
+    #[tokio::test]
+    async fn test_slot_limiting_blocks_excess_requests() {
+        // Tests the slot limiting mechanism:
+        // - Creates a ValidatorManager with max_requests_per_node = 2
+        // - Starts two slow requests that acquire both available slots
+        // - Starts a third request and verifies it's blocked waiting for a slot (execution count stays at 2)
+        // - Completes the first request to release a slot
+        // - Verifies the third request now acquires the freed slot and executes (execution count becomes 3)
+        // - Confirms all requests complete successfully
+        use linera_base::identifiers::BlobType;
```

Comment:
Should probably be at the start of the file.

Reply:
It's used in this single test 🤔